pkg/scheduler/framework/v1alpha1/metrics_recorder.go

```diff
@@ -0,0 +1,119 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"time"
+
+	k8smetrics "k8s.io/component-base/metrics"
+	"k8s.io/kubernetes/pkg/scheduler/metrics"
+)
+
+type metricsRecorder interface {
+	observeExtensionPointDurationAsync(extensionPoint string, status *Status, value float64)
+	observePluginDurationAsync(extensionPoint, pluginName string, status *Status, value float64)
+}
+
+type noopMetricsRecorder struct{}
+
+func (*noopMetricsRecorder) observeExtensionPointDurationAsync(extensionPoint string, status *Status, value float64) {
+}
+func (*noopMetricsRecorder) observePluginDurationAsync(extensionPoint, pluginName string, status *Status, value float64) {
+}
+
+// frameworkMetric is the data structure passed in the buffer channel between the main framework thread
+// and the metricsRecorder goroutine.
+type frameworkMetric struct {
+	metric      *k8smetrics.HistogramVec
+	labelValues []string
+	value       float64
+}
+
+// metricRecorder records framework metrics in a separate goroutine to avoid overhead in the critical path.
+type realMetricsRecorder struct {
+	// bufferCh is a channel that serves as a metrics buffer before the metricsRecorder goroutine reports it.
+	bufferCh chan *frameworkMetric
+	// if bufferSize is reached, incoming metrics will be discarded.
+	bufferSize int
+	// how often the recorder runs to flush the metrics.
+	interval time.Duration
+
+	stopCh      chan struct{}
+	isStoppedCh chan struct{}
+}
+
+func newMetricsRecorder(bufferSize int, interval time.Duration) *realMetricsRecorder {
+	recorder := &realMetricsRecorder{
+		bufferCh:    make(chan *frameworkMetric, bufferSize),
+		bufferSize:  bufferSize,
+		interval:    interval,
+		stopCh:      make(chan struct{}),
+		isStoppedCh: make(chan struct{}),
+	}
+	go recorder.run()
+	return recorder
+}
+
+func (r *realMetricsRecorder) observeExtensionPointDurationAsync(extensionPoint string, status *Status, value float64) {
```

Comment:
can you add a comment explaining the difference between this and the function below.